proposals/0360-opaque-result-types-with-availability.md

```diff
@@ -87,18 +87,139 @@ This is a dead-end for the library author although SE-0244 states that it should
 
 ## Proposed solution
 
-To bridge this usability gap I propose to relax same-type restriction for `return`s inside of availability conditions as follows:
+To bridge this usability gap, I propose to relax the same-type restriction for functions with `if #available` conditions as follows:
 
-A function returning an opaque result type is allowed to return values of different concrete types from conditionally available `if #available`  branches without any other dynamic conditions, if and only if, all universally available `return` statements in its body return a value of the same concrete type `T`. All returned values regardless of their location must meet all of the constraints stated on the opaque type.
+An *unconditional availability clause* is an `if` or `else if` clause that satisfies the following conditions:
+
+  - The clause is part of an `if` statement at the top level of the containing function.
+  - There are no `return` statements in the containing function prior to the `if` statement.
+  - The condition of the clause is an `#available` condition.
+  - The clause is either the initial `if` clause or an `else if` clause immediately following an unconditional availability clause.
+  - The clause contains at least one `return` statement.
+  - All paths through the block controlled by the clause terminate by either returning or throwing.
+
+1. `if #available` is a top-level statement in a function and *is not* preceded by any block that can return.
+
+  ```swift
+  func test() -> some Shape {
+    if #available(macOS 100, *) { ✅
+      return Rectangle()
+    }
+
+    return self
+  }
+  ```
+  
+  The following is incorrect because `if #available` is preceded by a dynamic condition that returns:
+  
+  ```swift
+  func test() -> some Shape {
+    guard let x = <opt-value> else {
+      return ...
+    }
+    
+    if #available(macOS 100, *) { ❌
+      return Rectangle()
+    }
+
+    return self
+  }
+  ```
+
+  Similarly, the following is incorrect because `if #available` appears inside of a loop:
+
+  ```swift
+  func test() -> some Shape {
+    for ... {
+      if #available(macOS 100, *) { ❌
+        return Rectangle()
+      }
+    }
+    return self
+  }
+  ```
+
+2. All `return` statements outside of unconditional availability clauses must return the same type as each other, and this type must be as available as the containing function.
```

Comment:
We're transitioning back into rules.  I think that's clear enough just because there are no immediate examples.  This should not be part of a list, however.